ecosystem/typescript/sdk/src/transaction_builder/builder.ts

```diff
@@ -292,3 +295,98 @@ export class TransactionBuilderABI {
     throw new Error("Invalid ABI.");
   }
 }
+
+export type RemoteABIBuilderConfig = Partial<Omit<ABIBuilderConfig, "sender">> & {
+  sender: MaybeHexString | AccountAddress;
+};
+
+interface AptosClientInterface {
+  getAccountModules: (arg0: string) => Promise<Gen.MoveModuleBytecode[]>;
+  getAccount: (accountAddress: MaybeHexString) => Promise<Gen.AccountData>;
+  getChainId: () => Promise<number>;
+}
+
+/**
+ * This transaction builder downloads JSON ABIs from the fullnodes.
+ * It then translates the JSON ABIs to the format that is accepted by TransactionBuilderABI
+ */
+export class TransactionBuilderRemoteABI {
+  // We don't want the builder to depend on the actual AptosClient. There might be circular dependencies.
+  constructor(
+    private readonly aptosClient: AptosClientInterface,
+    private readonly builderConfig: RemoteABIBuilderConfig,
+  ) {}
+
+  // Cache for 10 minutes
+  @MemoizeExpiring(10 * 60 * 1000)
+  async fetchABI(addr: string) {
+    const modules = await this.aptosClient.getAccountModules(HexString.ensure(addr).hex());
+    const abis = modules
+      .map((module) => module.abi)
+      .flatMap((abi) =>
+        abi.exposed_functions
+          .filter((ef) => ef.is_entry)
+          .map(
+            (ef) =>
+              ({
+                fullName: `${abi.address}::${abi.name}::${ef.name}`,
+                ...ef,
+              } as Gen.MoveFunction & { fullName: string }),
+          ),
+      );
+
+    const abiMap = new Map<string, Gen.MoveFunction & { fullName: string }>();
+    abis.forEach((abi) => {
+      abiMap.set(abi.fullName, abi);
+    });
+
+    return abiMap;
+  }
+
+  // Only support entry function payloads
+  async build(func: string, ty_tags: string[], args: any[]): Promise<RawTransaction> {
```

Comment:
I think we have a type alias for func, let's use that? Like MoveModuleFunctionId or something.

Reply:
if it is a type alias, then fine. I don't want people to pass in a struct.